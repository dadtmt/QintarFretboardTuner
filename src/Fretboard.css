:root {
    --background-black-fret: #246B61;
    --color-black-fret: #6BA099;
    --background-white-fret: #AA6839;
    --color-white-fret: #FFCDAA;
  }
  
  main {
    display: flex;
    justify-content: space-around;
    font-size: 1.2rem;
  }
  
  ol {
    width: 100%;
    height: 100vh;
    display: flex;
    flex-direction: column;
    margin: 0;
    padding: 0;
  }
  
  main>ol:nth-child(1) {
    width: 5%;  
  }
  
  main>ol:nth-child(1) li {
    text-align: center;
    height: 15vh;
  }
  
  li>ol {
    height: 15vh;
    flex-direction: row;  
  }
  
  li {
    list-style-type: none;
  }
  
  li li {
    width: 100%;
    text-align: center;
    padding-top: 0.2rem;
  }
  
  li:nth-child(odd) li:nth-child(odd) {
    background: var(--background-white-fret);
    color: var(--color-white-fret);
    
  }
  li:nth-child(odd) li:nth-child(even) {
    background: var(--background-black-fret);
    color: var(--color-black-fret);
    
  }
  li:nth-child(even) li:nth-child(even) {
    background: var(--background-white-fret);
    color: var(--color-white-fret);
    
  }
  li:nth-child(even) li:nth-child(odd) {
    background: var(--background-black-fret);
    color: var(--color-black-fret);
    
  }
  
  li:nth-child(even) {
    background: #2D4671;
    color: #7689A9;
  }
  
  li:nth-child(odd) {
    background: #408E2F;
    color: #9AD58E;
  }

  li li.selected {
    background: rgb(192, 228, 198) !important; 
    color: black!important;
  }