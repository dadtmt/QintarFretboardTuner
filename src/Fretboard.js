import React, { Fragment, useState } from 'react'
import { simplify } from '@tonaljs/note'
import { chord } from '@tonaljs/chord'
import classNames from 'classnames'
import { generateFretboard } from './generators'
import './Fretboard.css'

function TonalInput({
  initialValue,
  label,
  getTonal,
  setTonal,
  tonalType,
  validTonal
}) {
  const [tonalError, setTonalError] = useState(false)
  const [inputValue, setInputValue] = useState(initialValue)
  const handleValueChange = value => {
    const wantedTonalValue = getTonal(value)
    if (validTonal(wantedTonalValue)) {
      setTonalError(false)
      setTonal(wantedTonalValue)
    } else {
      setTonalError(true)
    }
    setInputValue(value)
  }

  return (
    <label htmlFor={label}>
      {`${label} : `}
      <input
        type="text"
        value={inputValue}
        onChange={e => handleValueChange(e.target.value)}
      />
      {tonalError && <span>{`not a valid ${tonalType}`}</span>}
    </label>
  )
}

function Fretboard() {
  const initialDeepestNote = 'F1'
  const initialChord = 'C3maj7'

  const [fretboard, setFretboard] = useState(
    generateFretboard(initialDeepestNote)
  )
  const [selectedChord, setSelectedChord] = useState(initialChord)

  return (
    <Fragment>
      <form action="">
        <TonalInput
          initialValue={initialDeepestNote}
          label="Deepest note"
          getTonal={simplify}
          setTonal={noteValue => setFretboard(generateFretboard(noteValue))}
          tonalType="note"
          validTonal={noteValue => noteValue !== ''}
        />
        <TonalInput
          initialValue={initialChord}
          label="Selected Chord"
          getTonal={chord}
          setTonal={chordValue => setSelectedChord(chordValue.name)}
          tonalType="chord"
          validTonal={chordValue => !chordValue.empty}
        />
      </form>
      <main>
        <ol>
          {fretboard.map(guitarString => (
            <li key={guitarString.flatNote}>{guitarString.flatNote}</li>
          ))}
        </ol>
        <ol>
          {fretboard.map(guitarString => (
            <li key={guitarString.flatNote}>
              <ol>
                {guitarString.frets.map(fret => (
                  <li
                    key={fret}
                    className={classNames({
                      selected: chord(selectedChord).notes.includes(fret)
                    })}>
                    {fret}
                  </li>
                ))}
              </ol>
            </li>
          ))}
        </ol>
      </main>
    </Fragment>
  )
}

export default Fretboard
