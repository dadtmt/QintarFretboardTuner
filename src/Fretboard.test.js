import React from 'react'
import ReactDOM from 'react-dom'
import Fretboard from './Fretboard'
import renderer, { act } from 'react-test-renderer'
import { toMatchDiffSnapshot } from 'snapshot-diff'

expect.extend({ toMatchDiffSnapshot })

it('renders correctly', () => {
  const component = renderer.create(<Fretboard />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  const wrongEventValue = { target: { value: 'xVfGu' } }
  act(() => {
    component.root.findAllByType('input')[0].props.onChange(wrongEventValue)
  })

  const treeUpdate = component.toJSON()
  console.log(treeUpdate)
  expect(tree).toMatchDiffSnapshot(treeUpdate)
})

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Fretboard />, div)
  ReactDOM.unmountComponentAtNode(div)
})
