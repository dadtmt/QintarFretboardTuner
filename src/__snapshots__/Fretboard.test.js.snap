// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <form
    action=""
  >
    <label
      htmlFor="Deepest note"
    >
      Deepest note : 
      <input
        onChange={[Function]}
        type="text"
        value="F1"
      />
    </label>
    <span>
      Click on a fret to display major chords
    </span>
  </form>,
  <main>
    <ol>
      <li>
        <ol>
          <li
            onClick={[Function]}
          >
            <aside>
              0
            </aside>
            <article>
              E4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              1
            </aside>
            <article>
              F4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              2
            </aside>
            <article>
              Gb4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              3
            </aside>
            <article>
              G4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              4
            </aside>
            <article>
              Ab4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              5
            </aside>
            <article>
              A4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              6
            </aside>
            <article>
              A#4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              7
            </aside>
            <article>
              B4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              8
            </aside>
            <article>
              C5
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              9
            </aside>
            <article>
              Db5
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              10
            </aside>
            <article>
              D5
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              11
            </aside>
            <article>
              Eb5
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              12
            </aside>
            <article>
              E5
            </article>
            <div>
              
            </div>
          </li>
        </ol>
      </li>
      <li>
        <ol>
          <li
            onClick={[Function]}
          >
            <aside>
              13
            </aside>
            <article>
              A3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              14
            </aside>
            <article>
              A#3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              15
            </aside>
            <article>
              B3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              16
            </aside>
            <article>
              C4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              17
            </aside>
            <article>
              Db4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              18
            </aside>
            <article>
              D4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              19
            </aside>
            <article>
              D#4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              20
            </aside>
            <article>
              E4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              21
            </aside>
            <article>
              F4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              22
            </aside>
            <article>
              Gb4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              23
            </aside>
            <article>
              G4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              24
            </aside>
            <article>
              Ab4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              25
            </aside>
            <article>
              A4
            </article>
            <div>
              
            </div>
          </li>
        </ol>
      </li>
      <li>
        <ol>
          <li
            onClick={[Function]}
          >
            <aside>
              26
            </aside>
            <article>
              D3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              27
            </aside>
            <article>
              D#3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              28
            </aside>
            <article>
              E3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              29
            </aside>
            <article>
              F3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              30
            </aside>
            <article>
              Gb3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              31
            </aside>
            <article>
              G3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              32
            </aside>
            <article>
              G#3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              33
            </aside>
            <article>
              A3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              34
            </aside>
            <article>
              A#3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              35
            </aside>
            <article>
              B3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              36
            </aside>
            <article>
              C4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              37
            </aside>
            <article>
              Db4
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              38
            </aside>
            <article>
              D4
            </article>
            <div>
              
            </div>
          </li>
        </ol>
      </li>
      <li>
        <ol>
          <li
            onClick={[Function]}
          >
            <aside>
              39
            </aside>
            <article>
              G2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              40
            </aside>
            <article>
              G#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              41
            </aside>
            <article>
              A2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              42
            </aside>
            <article>
              A#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              43
            </aside>
            <article>
              B2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              44
            </aside>
            <article>
              C3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              45
            </aside>
            <article>
              C#3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              46
            </aside>
            <article>
              D3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              47
            </aside>
            <article>
              D#3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              48
            </aside>
            <article>
              E3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              49
            </aside>
            <article>
              F3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              50
            </aside>
            <article>
              Gb3
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              51
            </aside>
            <article>
              G3
            </article>
            <div>
              
            </div>
          </li>
        </ol>
      </li>
      <li>
        <ol>
          <li
            onClick={[Function]}
          >
            <aside>
              52
            </aside>
            <article>
              C2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              53
            </aside>
            <article>
              C#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              54
            </aside>
            <article>
              D2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              55
            </aside>
            <article>
              D#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              56
            </aside>
            <article>
              E2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              57
            </aside>
            <article>
              F2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              58
            </aside>
            <article>
              F#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              59
            </aside>
            <article>
              G2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              60
            </aside>
            <article>
              G#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              61
            </aside>
            <article>
              A2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              62
            </aside>
            <article>
              A#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              63
            </aside>
            <article>
              B2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              64
            </aside>
            <article>
              C3
            </article>
            <div>
              
            </div>
          </li>
        </ol>
      </li>
      <li>
        <ol>
          <li
            onClick={[Function]}
          >
            <aside>
              65
            </aside>
            <article>
              F1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              66
            </aside>
            <article>
              F#1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              67
            </aside>
            <article>
              G1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              68
            </aside>
            <article>
              G#1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              69
            </aside>
            <article>
              A1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              70
            </aside>
            <article>
              A#1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              71
            </aside>
            <article>
              B1
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              72
            </aside>
            <article>
              C2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              73
            </aside>
            <article>
              C#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              74
            </aside>
            <article>
              D2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              75
            </aside>
            <article>
              D#2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              76
            </aside>
            <article>
              E2
            </article>
            <div>
              
            </div>
          </li>
          <li
            onClick={[Function]}
          >
            <aside>
              77
            </aside>
            <article>
              F2
            </article>
            <div>
              
            </div>
          </li>
        </ol>
      </li>
    </ol>
  </main>,
]
`;

exports[`renders correctly 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -7,12 +7,15 @@
      >
        Deepest note : 
        <input
          onChange={[Function onChange]}
          type=\\"text\\"
-         value=\\"F1\\"
+         value=\\"xVfGu\\"
        />
+       <span>
+         not a valid note
+       </span>
      </label>
      <span>
        Click on a fret to display major chords
      </span>
    </form>,"
`;
